compiler/src/iree/compiler/Dialect/Flow/Transforms/InterchangeTransposeGenericOps.cpp

```diff
@@ -0,0 +1,105 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+//===--------------- InterchangeTransposeGenericOps.cpp -------------------===//
+//
+// Interchange loops in generic ops to make the transpose happen on the outputs
+// instead of inputs.
+//
+//===----------------------------------------------------------------------===//
+
+#include "iree/compiler/Dialect/Flow/Transforms/FusionUtils.h"
+#include "iree/compiler/Dialect/Flow/Transforms/PassDetail.h"
+#include "iree/compiler/Dialect/Flow/Transforms/Passes.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+namespace {
+
+struct TransposeGenericOpPattern : public OpRewritePattern<linalg::GenericOp> {
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    if (genericOp.getNumInputs() == 0 || genericOp.getNumOutputs() == 0)
+      return failure();
+
+    // check the generic op can be fused with the input operands.
+    auto indexingMap0 =
+        genericOp.getTiedIndexingMap(genericOp.getInputOperand(0));
+    if (indexingMap0.isIdentity()) {
+      // already in a good form.
+      return failure();
+    }
+
+    if (indexingMap0.getNumDims() != genericOp.getNumParallelLoops())
+      return failure();
+
+    for (OpOperand *operand : genericOp.getInputOperands()) {
+      if (!areLinalgOpsFusableUsingTileAndFuse(*operand,
```

Comment:
We may just need to normalize genericOps to have the identity indexing map if it does not egardless of the producer. Do you see any side effect?

Reply:
It can get into a loop, where you switch it w.r.t one producer and switch it again w.r.t to another. So need a check here that make sure the pattern applies only once. Checking for the first named op producer and making that identity should address this

Reply:
I pushed a change to normalize the input indexing maps to identity. Thus, once the input indexing maps are identity, the pass does not revert so we won't have a loop.